### PR TITLE
Support IPv6 raw sockets

### PR DESCRIPTION
**Pull request by dlrobertson**
- Add support for IPv6 raw sockets.
 - Update tests.

### CODE REVIEW

**Comment by dlrobertson** [comment]
> The very first PR should probably just handle IPv6 packets and forward them to the socket. The second should probably handle sending packets via RawSocket.

Flipped the order of these.

File: src/socket/raw.rs [hunk 1]
```diff
@@ -205,7 +207,13 @@ impl<'a, 'b> RawSocket<'a, 'b> {
                     Ok((IpRepr::Ipv4(ipv4_repr), packet.payload()))
                 }
                 #[cfg(feature = "proto-ipv6")]
-                IpVersion::Ipv6 => Err(Error::Unrecognized),
+                IpVersion::Ipv6 => {
+                    let mut packet = Ipv6Packet::new_checked(buffer.as_mut())?;
+                    if packet.next_header() != protocol { return Err(Error::Unaddressable) }
+                    let packet = Ipv6Packet::new(&*packet.into_inner());
+                    let ipv4_repr = Ipv6Repr::parse(&packet)?;
```
Comment (whitequark) [inline src/socket/raw.rs]: `ipv4_repr` → `ipv6_repr`

File: src/socket/raw.rs [hunk 2]
```diff
@@ -205,7 +207,13 @@ impl<'a, 'b> RawSocket<'a, 'b> {
                     Ok((IpRepr::Ipv4(ipv4_repr), packet.payload()))
                 }
                 #[cfg(feature = "proto-ipv6")]
-                IpVersion::Ipv6 => Err(Error::Unrecognized),
+                IpVersion::Ipv6 => {
+                    let mut packet = Ipv6Packet::new_checked(buffer.as_mut())?;
+                    if packet.next_header() != protocol { return Err(Error::Unaddressable) }
```
Comment (whitequark) [inline src/socket/raw.rs]: You should fill the checksum.

File: src/socket/raw.rs [hunk 3]
```diff
@@ -291,123 +302,238 @@ mod test {
...
+        #[cfg(feature = "proto-ipv4")]
+        {
+            let socket = match RawSocket::new(IpVersion::Ipv4,
+                                              IpProtocol::Unknown(ipv4_locals::IP_PROTO+1),
+                                              buffer(1), buffer(1)) {
+                Socket::Raw(socket) => socket,
+                _ => unreachable!()
+            };
+            assert!(!socket.accepts(&ipv4_locals::HEADER_REPR));
+            #[cfg(feature = "proto-ipv6")]
```
Comment (whitequark) [inline src/socket/raw.rs]: Useless `#[cfg]`.

File: src/socket/raw.rs [hunk 4]
```diff
@@ -291,123 +302,238 @@ mod test {
...
+        #[cfg(feature = "proto-ipv6")]
+        {
+            let socket = match RawSocket::new(IpVersion::Ipv6,
+                                              IpProtocol::Unknown(ipv6_locals::IP_PROTO+1),
+                                              buffer(1), buffer(1)) {
+                Socket::Raw(socket) => socket,
+                _ => unreachable!()
+            };
+            assert!(!socket.accepts(&ipv6_locals::HEADER_REPR));
+            #[cfg(feature = "proto-ipv4")]
```
Comment (whitequark) [inline src/socket/raw.rs]: As above.

File: src/socket/raw.rs (hunk 2 above)
Comment (dlrobertson) [reply src/socket/raw.rs]: IPv6 packets don't have a checksum. I assumed that the user takes care of the upper layer. Is that not correct?

File: src/socket/raw.rs (hunk 3 above)
Comment (dlrobertson) [reply src/socket/raw.rs]: I use an ipv6 specific variable below and this block is run if `proto-ipv4` is enabled, so if `proto-ipv4` is enabled but `proto-ipv6` is not you'll get a compile failure.

File: src/socket/raw.rs (hunk 4 above)
Comment (dlrobertson) [reply src/socket/raw.rs]: Same

File: src/socket/raw.rs (hunk 1 above)
Comment (dlrobertson) [reply src/socket/raw.rs]: Updated

File: src/socket/raw.rs (hunk 3 above)
Comment (whitequark) [reply src/socket/raw.rs]: Oh, I misread.

**Pull request merged**